docs/src/main/asciidoc/credentials-provider.adoc

```diff
@@ -0,0 +1,194 @@
+An alternative is to define both username and password in Vault and drop the `quarkus.datasource.username`
+property from configuration. All consuming extensions do support the ability to fetch both the username
+and password from the provider, or just the password.
+
+== Custom Credentials Provider
+
+Implementing a custom credentials provider is the only option when a vault product is not yet supported in Quarkus, or if credentials need to be retrieved from a custom store.
+
+The only interface to implement is:
+
+[source, java]
+----
+public interface CredentialsProvider {
+
+    String USER_PROPERTY_NAME = "user";
+    String PASSWORD_PROPERTY_NAME = "password";
+
+    Map<String, String> getCredentials(String credentialsProviderName);
+
+}
+----
+
+`USER_PROPERTY_NAME` and `PASSWORD_PROPERTY_NAME` are standard properties that should be recognized by any consuming extension that support username/password based authentication.
+
+Here is a simple implementation example:
+
+[source, java]
+----
+@ApplicationScoped
+@Unremovable
+public class MyCredentialsProvider implements CredentialsProvider {
+
+	@Override
+	public Map<String, String> getCredentials(String credentialsProviderName) {
+
+		Map<String, String> properties = new HashMap<>();
+		properties.put(USER_PROPERTY_NAME, "hibernate_orm_test");
+		properties.put(PASSWORD_PROPERTY_NAME, "hibernate_orm_test");
+		return properties;
+	}
+
+}
+----
+
+Note that we decided here to return both the username and the password.
+
+This provider may be used in a datasource definition like this:
+
+[source, properties]
+----
+quarkus.datasource.db-kind=postgresql
+quarkus.datasource.credentials-provider=custom
+quarkus.datasource.jdbc.url=jdbc:postgresql://localhost:5431/hibernate_orm_test
+----
+
+It is also possible to pass configuration properties to the provider using standard MicroProfile `Config` injection:
+
+[source, properties]
+----
+custom.foo=bar
+----
+
+And in the provider implementation:
+
+[source, java]
+----
+@Inject
+Config config;
+
+@Override
+public Map<String, String> getCredentials(String credentialsProviderName) {
+
+    System.out.println("MyCredentialsProvider called with foo=" + config.getValue(credentialsProviderName + ".foo", String.class));
+    ...
+----
+
+== New Credentials Provider extension
+
+When creating a custom `Credentials Provider` in a new extension, there are a few additional considerations.
+
+First, you need to name it to avoid collisions in case multiple credentials providers are available in the project:
+
+[source, java]
+----
+@ApplicationScoped
+@Unremovable
+@Named("my-credentials-provider")
+public class MyCredentialsProvider implements CredentialsProvider {
+----
+
+It is the responsibility of the consumer to allow a `credentials-provider-type` property:
+
+[source, properties]
+----
+quarkus.datasource.credentials-provider = custom
+quarkus.datasource.credentials-provider-type = my-credentials-provider
```

Comment:
It might be too late now to change this, but this seems inconsistent. The value of `credentials-provider-type` is the CDI **name**, where I would expect that it would either be the class, or the property be named `credentials-provider-name`

Reply:
good point.
to me, the name is `mydatabase`, as in:
```
quarkus.vault.credentials-provider.mydatabase.kv-path=myapps/vault-quickstart/db

quarkus.datasource.credentials-provider = mydatabase
```

and reflected in the method parameter:
```
Map<String, String> getCredentials(String credentialsProviderName);
```

instead of `credentials-provider-type` or `credentials-provider-name`, may be `credentials-provider-cdi-name` or `credentials-provider-bean-name` would be more appropriate.

or we replace the name by an `id`:
```
Map<String, String> getCredentials(String credentialsProviderId);
```
and use `credentials-provider-name` as you suggest in place of `credentials-provider-type`.

as far as being too late, I suppose you are talking about 1.5? it is an easy change, and I would be surprised if anybody was using the `credentials-provider-type` property today. so we can do it now, or I can schedule this small evolution for 1.6. Let me know.

cc @gsmet

Reply:
I am personally in favor of making the change now if we have the time, but I'll let @gsmet decide

Reply:
Let's go with `credentials-provider-bean-name` but it needs to happen fast if we want it for 1.5 as I will cut the release at the beginning of the afternoon if all goes well.

Reply:
`credentials-provider-bean-name` sounds a bit too technical (will it always be a bean ?), `credentials-provider-name` seems a good compromise

Reply:
Looking at the code it seems to always be a bean name.

I am fine with the compromise suggestion

Reply:
@gsmet I am just seeing your message now. I think the renaming is too short since there are code changes. I would suggest that you merge the doc PR, and I am creating now a different PR for the naming change, that you may decide to apply on 1.5 or not.